kernel/net/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use irq_safety::MutexIrqSafe;
+use smoltcp::wire::Ipv4Address;
+use spin::Mutex;
+
+mod device;
+mod error;
+mod interface;
+
+pub use device::{Device, DeviceCapabilities};
+pub use error::{Error, Result};
+pub use interface::{Interface, IpAddress, IpCidr, SocketSet};
+pub use smoltcp::{phy, socket, time::Instant, wire};
+
+/// A randomly chosen IP address that must be outside of the DHCP range.
+///
+/// The default QEMU user-slirp network gives IP address of `10.0.2.*`.
+const DEFAULT_LOCAL_IP: &str = "10.0.2.15/24";
+
+/// Standard home router address.
+///
+/// `10.0.2.2` is the default QEMU user-slirp networking gateway IP.
+const DEFAULT_GATEWAY_IP: IpAddress = IpAddress::Ipv4(Ipv4Address::new(10, 0, 2, 2));
+
+// TODO: Make mutex rwlock?
+// TODO: Use atomic append-only vec?
+static NETWORK_INTERFACES: Mutex<Vec<Interface>> = Mutex::new(Vec::new());
```

Comment:
wrapping the inner registered interfaces in an Arc will relieve the multi-locking contention issues too.

Reply:
I'm not sure what the contention issues are. Cloning an `Interface` is basically cloning an `Arc` because its two fields are an `Arc` and a reference. I'm not sure wrapping it in an additional `Arc` will help.

Having said that I'd rather use an `Arc<Interface>>`, rather than have `interface.inner` be an `Arc` as currently, it isn't immediately clear that `Interface` is essentially a reference.